iocore/eventsystem/UnixEventProcessor.cc

```diff
@@ -265,17 +265,32 @@ EventProcessor::shutdown()
 Event *
 EventProcessor::spawn_thread(Continuation *cont, const char *thr_name, size_t stacksize)
 {
-  ink_release_assert(n_dthreads < MAX_EVENT_THREADS);
-  int n = ink_atomic_increment((int *)&n_dthreads, 1);
-  ink_release_assert(n < MAX_EVENT_THREADS);
-
+  /* Spawning threads in a live system - There are two potential race conditions in this logic. The
+     first is multiple calls to this method.  In that case @a all_dthreads can end up in a bad state
+     as the same entry is overwritten while another is left unitialized.
+     The other is read/write contention where another thread (e.g. the stats collection thread) is
+     iterating over the threads while the active count (@a n_dthreads) is being updated causing use
+     of not yet initialized array element.
+     This logic covers both situations. For write/write the actual array update is locked. The
+     potentially expensive set up is done outside the lock making the time spent locked small For
+     read/write it suffices to do the active count increment after initializing the array
+     element. It's not a problem if, for one cycle, a new thread is skipped.
+  */
+
+  // Do as much as possible outside the lock. Until the array element and count is changed
+  // this is thread safe.
   Event *e = eventAllocator.alloc();
 
   e->init(cont, 0, 0);
-  all_dthreads[n] = new EThread(DEDICATED, e);
-  e->ethread      = all_dthreads[n];
-  e->mutex        = all_dthreads[n]->mutex;
-  cont->mutex     = all_dthreads[n]->mutex;
+  e->ethread      = new EThread(DEDICATED, e);
+  e->mutex        = e->ethread->mutex;
+  cont->mutex     = e->ethread->mutex;
+  {
+    ink_mutex_acquire(&dedicated_spawn_thread_mutex);
+    ink_release_assert(n_dthreads < MAX_EVENT_THREADS);
+    all_dthreads[n_dthreads++] = new EThread(DEDICATED, e);
```

Comment:
No! This needs to be `e->thread`.
Also spoke with @dragon512 and he says I can't depend on the compiler doing the increment of `n_dthreads` before the array assignment, which is of course critical.